isis/src/base/apps/framestitch/framestitch.cpp

```diff
@@ -0,0 +1,221 @@
+#include "framestitch.h"
+
+#include <QString>
+
+#include "Brick.h"
+#include "Buffer.h"
+#include "Cube.h"
+#include "CubeAttribute.h"
+#include "Camera.h"
+#include "IException.h"
+#include "IString.h"
+#include "LineManager.h"
+#include "ProcessByBrick.h"
+#include "PvlGroup.h"
+#include "PvlKeyword.h"
+#include "Statistics.h"
+#include "UserInterface.h"
+
+using namespace std;
+using namespace Isis;
+
+namespace Isis {
+
+  /**
+   * Helper function to count the height of frames in a cube.
+   *
+   * Iterates over the lines in the cube and counts sequential all NULL lines
+   * assuming these are frames that have been removed from the cube.
+   *
+   * If different frame heights are computed within the cube or there are
+   * no all null lines in the cube an error is raised.
+   */
+  int computeFrameHeight(Cube *cube) {
+    // Use a Statistics object to track min/max height count
+    Statistics frameHeights;
+    LineManager cubeLine(*cube);
+    int currentFrameHeight = 0;
+    for (int line = 1; line <= cube->lineCount(); line++) {
+      cubeLine.SetLine(line);
+      cube->read(cubeLine);
+      Statistics lineStats;
+      lineStats.AddData(cubeLine.DoubleBuffer(), cubeLine.size());
+      // If the line is all NULL add it to the current frame count
+      if (lineStats.TotalPixels() == lineStats.NullPixels()) {
+        ++currentFrameHeight;
+      }
+      // If the line has non-NULL pixels and we have previously counted
+      // some all null lines add the count to our histogram and reset
+      else if (currentFrameHeight > 0) {
+        frameHeights.AddData(currentFrameHeight);
+        currentFrameHeight = 0;
+      }
+    }
+
+    // If the last line is part of a NULL frame handle it now
+    if (currentFrameHeight > 0) {
+      frameHeights.AddData(currentFrameHeight);
+    }
+
+    if (frameHeights.TotalPixels() == 0) {
+      QString msg = "Failed to find any NULL frames in cube [" + cube->fileName() + "]."
+                    "Please manually enter the frame height.";
+      throw IException(IException::User, msg, _FILEINFO_);
+    }
+    if (frameHeights.Minimum() != frameHeights.Maximum()) {
+      QString msg = "Found different frame heights between [" + toString((int)frameHeights.Minimum())
+                  + "] and [" + toString((int)frameHeights.Maximum()) + "] lines in cube ["
+                  + cube->fileName() + "]. Please manually enter the frame height.";
+      throw IException(IException::User, msg, _FILEINFO_);
+    }
+
+    // Use the median to handle any outliers
+    return frameHeights.Average();
```

Comment:
Should this be average or median?

Reply:
I switched to a statistics object instead of a histogram and forgot to update the comment. It doesn't matter because there should only be 1 value accumulated by the time we get here